### ISSUE TITLE
Configurable API templates in Apicurio Studio

### ISSUE THREAD

**Issue by DyspC** [Apicurio/apicurio-studio#1562]
Hello, 

We're deploying apicurio to edit both OAS & AAI APIs and we'd like to give our users custom templates with for example our kafka brokers or imports of our message traits / shared error payloads.

Rebuilding docker images from source to allow custom templates does not feel right so we'll work on a way to load them later instead of hardcoding them.

We've discussed a bit those two ideas are sticking:
- Extending the docker image to copy a template folder with all metadata and API content
- Adding a configuration key like apicurio-ui.templates.url that points to an array like [this one](https://github.com/lbroudoux/apicurio-studio/blob/fb8d0b6974258c2e1359ee28fdce98b7653356da/front-end/studio/src/app/services/template.service.ts#L36)

Both solutions will probably need an additional flag to state whether the custom templates should replace the default ones or extend the selection and serving the templates through one of the backends would allow using some cache.

I don't think we are the only ones with this specific need so we should discuss the best way to handle this and make it worth a PR.

I have a personal preference for the configuration keys because it does not force a parallel docker release train

**Comment by EricWittmann** [comment]
I like this idea very much.  I think we should completely remove the hard-coded array of templates from the UI layer and instead serve the list of them from the back-end.

I don't have a strong opinion between the template folder and the conf key (probably a preference for the config key for the reason already given).  Another option, however, would be to update the back-end to manage the collection of templates as a resource in the REST API.  This would allow dynamic configuration of these templates and management of them via an API (and eventually a UI).

In any case, I think the first step would be to at least expose a REST API endpoint that would allow the UI to query for the list of templates.  That API can (at first) be read-only and driven from config.  And we can then enhance it in the future to allow full dynamic management.

**Comment by DyspC** [comment]
The REST API approach is nice indeed. Although this calls for user roles and scoping (user, team, instance...) since we don't want anybody to create instance wide templates.

With a new `Template` ApiContentType that holds a reference to a `Document` row of the same design_id (only the version would be needed then) this could fit in the existing DDL.

This would work for user defined templates but will need a bit of rework when (if?) scopes are introduced (I think the json stored in a character varying column won't allow postgresql's JSONB operations and I don't know what mysql and h2 have to offer to avoid application side filtering)

I don't know how it can be translated for default templates though (forcing them into the DB with a fake user is hack-ish and will cause problems with upstream updates)

**Comment by EricWittmann** [comment]
Agreed that user scoping is needed for this - so only admins can add templates.  That's another reason to start off with a static list backed by an API and evolve over time to something fully managed.

**Comment by DyspC** [comment]
I think I'm getting close to what I was thinking about in my last message, what do you think about using snapshots of stored designs as templates?

![image](https://user-images.githubusercontent.com/17125930/131501420-653ecdcf-2323-4303-96ba-9d113139491d.png)

![image](https://user-images.githubusercontent.com/17125930/131501561-8689a940-b7d9-4304-991c-6b0d84091696.png)

![image](https://user-images.githubusercontent.com/17125930/131501763-f2ecfe41-8aec-4730-ac3a-467c6a12eeb1.png)

ATM I hid the functionnality behind an API/UI feature flag that -when enabled- allows all users to access the template publication page.

Submitting the form applies the pending commands on a design and creates a row like this in the api_content table. 

```
 design_id | version | type |                                                 data                                                 
         8 |      88 |    4 | {"designId":"8","contentVersion":65,"name":"MyTestTemplate","description":"A test template "}        
         8 |      91 |    4 | {"designId":"8","contentVersion":90,"name":"MyTestTemplate","description":"A test template "}        
         8 |      94 |    4 | {"designId":"8","contentVersion":93,"name":"MyTestTemplate2","description":""}                       
         3 |     101 |    4 | {"designId":"3","contentVersion":41,"name":"petstore20","description":null}                          
        19 |     124 |    4 | {"designId":"19","contentVersion":123,"name":"Sth Happens","description":"it does"}        
```

The latest row for each design is then dereferenced and served along with static templates by the /templates endpoint

This means an API can be promoted at any time and backups are kept in case the latest promotion is removed/revoked

Although I don't have any template ID to create a DELETE endpoint (unless /designs/:designId/templates/:contentVersion is acceptable) perhaps a new table like sharing would be a better fit for this purpose (uuids, index for api types without api_designs joins)

Perhaps the synthesis panel I put in the template promotion page should be moved to a single page independent of the parent design (or else it will become difficult to find where to go to update/remove a template)

What do you think about this?

**Comment by DyspC** [comment]
As a side note I've been told we could take a look at the user scoping part as well if you have some hints.

If you'd like I think discussing this in another issue would be better

**Comment by EricWittmann** [comment]
omg I'm so far behind on my github email notifications.  Sorry!

So I really like the user experience for creating a template.  The action in the UI is great.  This will make it so easy to create and edit the templates.

Also very interesting idea, using the content table to essentially store a snapshot.  I would need to think through the ramifications of that.  

Originally I was imagining a simple table.  A `templates` table with the following columns:

* templateId (varchar)
* name (varchar)
* description (varchar)
* owner (varchar)
* template (clob)

With a set of CRUD operations to manage the rows.  This allows for the UX you've designed (the action to promote an API design as a template) while also allowing templates to be added without first creating them as actual API designs.  

Now that I think of it, I would probably keep just about everything you've already done, but **also** have the extra table with the standard REST API.  So what we display in the UI when listing the templates just comes from the `templates` table, but the user experience of creating a template from an API design is exactly what you've created.

Templates created by users can not be global, so the only way to scope them **right now** is to do it per-user.  Similarly to how we do the validation profiles today.  We really need to get to work on Teams or something similar though.  At that point, the templates can be scoped per Team.

**Comment by DyspC** [comment]
> Originally I was imagining a simple table. A `templates` table with the following columns:
> 
> * templateId (varchar)
> * name (varchar)
> * description (varchar)
> * owner (varchar)
> * template (clob)

And perhaps an API type column so they do not need a parser to be filtered?

> With a set of CRUD operations to manage the rows. This allows for the UX you've designed (the action to promote an API design as a template) while also allowing templates to be added without first creating them as actual API designs.
> 
> Now that I think of it, I would probably keep just about everything you've already done, but **also** have the extra table with the standard REST API. So what we display in the UI when listing the templates just comes from the `templates` table, but the user experience of creating a template from an API design is exactly what you've created.

Do you mean having both a new table and the template rows in `api_content`? I'm not sure of how they can mix

> Templates created by users can not be global, so the only way to scope them **right now** is to do it per-user. Similarly to how we do the validation profiles today. We really need to get to work on Teams or something similar though. At that point, the templates can be scoped per Team.

I don't think the per-user scope on templates would have any benefit that isn't reachable right now with the clone API functionality, it's more restrictive than ACLs on the source document and being able to create a template the editor wouldn't accept in a ReplaceDocument command (cf. `while also allowing templates to be added without first creating them as actual API designs`) only means a fresh document could be invalid.

**Comment by EricWittmann** [comment]
> 
> 
> > Originally I was imagining a simple table. A `templates` table with the following columns:
> > 
> > * templateId (varchar)
> > * name (varchar)
> > * description (varchar)
> > * owner (varchar)
> > * template (clob)
> 
> And perhaps an API type column so they do not need a parser to be filtered?

Yes good point.  Also that.

> > With a set of CRUD operations to manage the rows. This allows for the UX you've designed (the action to promote an API design as a template) while also allowing templates to be added without first creating them as actual API designs.
> > Now that I think of it, I would probably keep just about everything you've already done, but **also** have the extra table with the standard REST API. So what we display in the UI when listing the templates just comes from the `templates` table, but the user experience of creating a template from an API design is exactly what you've created.
> 
> Do you mean having both a new table and the template rows in `api_content`? I'm not sure of how they can mix

I think they would mix similarly to how we currently handle Publish to Git.  When promoting the API design as a template, we would just use the normal CRUD API for managing templates.  But we would **also** add an entry to the content table to indicate that a Promote as Template operation was done.  This way it shows up in the activity log.  Very much like Publish to Git does today.

> > Templates created by users can not be global, so the only way to scope them **right now** is to do it per-user. Similarly to how we do the validation profiles today. We really need to get to work on Teams or something similar though. At that point, the templates can be scoped per Team.
> 
> I don't think the per-user scope on templates would have any benefit that isn't reachable right now with the clone API functionality, it's more restrictive than ACLs on the source document and being able to create a template the editor wouldn't accept in a ReplaceDocument command (cf. `while also allowing templates to be added without first creating them as actual API designs`) only means a fresh document could be invalid.

Point taken.  I don't necessarily disagree, but I can't allow the action to be available in a deployment like studio.apicur.io without some sort of scoping.  I can't let any user create a template that is then visible to all other users.  In corporate deployments that might be the desired result, but not in a shared environment like studio.apicur.io.  But perhaps (until we have some sort of organization based scoping) we could hide the feature behind a flag (it would complement the "share for everyone" global option).

Thoughts?

**Comment by DyspC** [comment]
> I can't allow the action to be available in a deployment like studio.apicur.io without some sort of scoping. I can't let any user create a template that is then visible to all other users. In corporate deployments that might be the desired result, but not in a shared environment like studio.apicur.io.

Oh indeed that's a lot of users. 
We're not using the share flag either out of staging environments but the way I thought about that was something along "We'll prepare the template on a preprod environment and manually insert the template reference in the main db". Since we do not need a whole lot of templates and very few people will be in charge of them it should do the trick.

I was recently POC-ing around the fact that keycloak can map the incoming identity provider claims to application roles (or the KC admin can assign roles directly to a user). 
The "can see and do everything" admin role would get out of hand real quick (especially with thousands of designs) but perhaps this approach can be introduced with the template publication.

**Comment by EricWittmann** [comment]
It would be a nice iterative improvement to have a way to identify a user as an Admin and give them access to this action.  Using Keycloak to map claims to roles would make a lot of sense.

**Comment by DyspC** [comment]
Alright then, I'll include your remarks into my branch and prepare a PR to have better discussion assets than screenshots.

Do you want the keycloak part stuffed in there or in another PR? I fear adding the role logic before any use case wouldn't make much sense.

**Comment by EricWittmann** [comment]
I'm thankful for the contributions so whatever is easiest for you.  If it doesn't make any difference to you I would slightly prefer the KC role stuff to be a separate PR.  But really, if that's a PITA then it's fine to put it all together.

**Comment by DyspC** [comment]
Ok, I'll prepare that part next week first and wait on the merge to remove the API template publication feature flag from the other branch
